### PR TITLE
layout: Account for transform in scrollable overflow

### LINKED ISSUE #36031: The translate element layout result is incorrect.
**Describe the bug:**
When an element is translated 50% to the left, the size of the layout it occupies is still in the position before translation.

**To Reproduce:**
They have different scrollWidth values on Firefox and Servo.

Image

testcase：
```
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <style id="style">
        #out {
            background-color: green;
        }
        #absolute {
            height: 1000px;
            width: 1000px;
            background-color: red;
            transform: translate(-50%, -50%);
        }
        body {
            background-color: pink;
        }

    </style>
</head>

<body class="">
<div id="out" class="">
    <div id="absolute" class="">
        absolute div
    </div>
</div>
</body>
<script>
    document.addEventListener("click", function (e) {
        document.body.appendChild(document.createTextNode(`clientWidth:${document.documentElement.clientWidth}, scrollWidth:${document.documentElement.scrollWidth}`));
    })
</script>
</html>
```

**Platform:**
ALL

### PR DESCRIPTION
**Pull request by stevennovaryo**
In the scrollable overflow calcutation, apply CSS transforms to boxes and scrollable overflow of the descendant. Clip unreachable scrollable overflow according to it's block start and inline start scrolling direction. And, renamed `Fragment::scrolling_overflow` to `Fragment::scrolling_overflow_for_parent` as it was calculating the scrolling overflow contribution from a child. 

Add several WPT tests, testing the transform interaction `rotate`, `scale`, and `skew` with scrollable overflow. There are several WPT test that are testing the interaction that not expected from current browsers implementation according to the spec.

Testing: Existing and new WPT. 
Fixes: #36031

### CODE REVIEW

File: components/layout_2020/style_ext.rs [hunk 1]
```diff
@@ -953,6 +964,24 @@ impl ComputedValuesExt for ComputedValues {
...
+        let writing_mode: WritingMode = self.writing_mode;
+        let inline_end_direction = writing_mode.inline_end_physical_side();
+        let block_end_direction = writing_mode.block_end_physical_side();
+
+        let rightward = inline_end_direction == PhysicalSide::Right ||
+            block_end_direction == PhysicalSide::Right;
+        let downward = inline_end_direction == PhysicalSide::Bottom ||
+            block_end_direction == PhysicalSide::Bottom;
+
+        // FIXME: We should consider the flex-container's CSS (e.g. flow-direction: column-reverse).
```
Comment (xiaochengh) [inline components/layout_2020/style_ext.rs:979]: Does Servo have a style guide about TODOs?

Anyway, it will be better to have assigned/tracked TODOs instead of anonymous ones.

For example: `TODO(stevennovaryo: We should consider blabla` or `TODO(#XXXXX): We should consider blabla`.

File: components/layout_2020/fragment_tree/box_fragment.rs [hunk 2]
```diff
@@ -238,30 +240,71 @@ impl BoxFragment {
...
+                        f32_rect_to_au_rect(transformed_overflow_box.to_untyped()).cast_unit();
+                }
+            }
+        }
+
+        overflow
+    }
+
+    /// <https://drafts.csswg.org/cssom-view/#scrolling-area>
+    /// We will "clip" the scrollable overflow based on it's overflow direction.
```
Comment (xiaochengh) [inline components/layout_2020/fragment_tree/box_fragment.rs:287]: it's --> its
Ditto below

File: components/layout_2020/fragment_tree/box_fragment.rs [hunk 3]
```diff
@@ -238,30 +240,71 @@ impl BoxFragment {
...
+    /// Specifically, a scrolling area would not consider area beyond it's block start and inline start corner.
```
Comment (xiaochengh) [inline components/layout_2020/fragment_tree/box_fragment.rs:288]:
```suggestion
    /// Specifically, a scrolling area would not consider area beyond it's block start and inline start edges.
```

File: components/layout_2020/style_ext.rs [hunk 4]
```diff
@@ -953,6 +964,24 @@ impl ComputedValuesExt for ComputedValues {
         };
         has_percentage(box_offsets.block_start) || has_percentage(box_offsets.block_end)
     }
+
+    // <https://drafts.csswg.org/cssom-view/#overflow-directions>
+    fn scrolling_overflow_direction(&self) -> ScrollingOverflowDirection {
```
Comment (xiaochengh) [inline components/layout_2020/style_ext.rs:969]: Can we simply name the method `overflow_directions`, and the struct `OverflowDirections`, to match the spec?

**Review by xiaochengh — COMMENTED** [review]
Nice! LGTM % some minor comments.

File: components/layout_2020/style_ext.rs (hunk 4 above)
Comment (stevennovaryo) [reply components/layout_2020/style_ext.rs]: Renamed, thanks!

File: components/layout_2020/style_ext.rs (hunk 1 above)
Comment (stevennovaryo) [reply components/layout_2020/style_ext.rs:979]: I am not sure. I suppose it is much better to make the appropriate issue if it's clear enough. But I would like to know the details of flex and writing mode progress first.

**Comment by stevennovaryo** [comment]
In the discussion with @xiaochengh, it seems that several part of current scrollable overflow implementation in browsers does not match the spec (i.e. https://github.com/servo/servo/issues/36031#issuecomment-2742593335). Following that, I have added several WPT test, testing several cases that was deemed unexpected.

File: components/layout_2020/style_ext.rs (hunk 1 above)
Comment (xiaochengh) [reply components/layout_2020/style_ext.rs:979]: Direction and writing mode define the "logical axes", which are different from the physical (i.e., x & y) axes. See the "coordinate spaces" section of:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/layout/README.md#coordinate-spaces

Note that direction and writing mode never result in a bottom-to-top block direction.

Similar effect can only be achieve with flex layout using `flow-direction: column-reverse`

File: tests/wpt/tests/css/css-overflow/scrollable-overflow-transform-005.html [hunk 5]
```diff
@@ -0,0 +1,68 @@
+<!DOCTYPE html>
+<meta charset="utf-8">
+<title>CSS Overflow: Scrollable Overflow Transform Initial Untransformed Box</title>
+<link rel="author" title="Jo Steven Novaryo" href="mailto:steven.novaryo@gmail.com">
+<link rel="help" href="https://drafts.csswg.org/css-overflow-3/#scrollable" />
+<meta name="assert" content="Checks that the initial untransformed box is not included in scrollable overflow.">
```
Comment (xiaochengh) [inline tests/wpt/tests/css/css-overflow/scrollable-overflow-transform-005.html:6]: What does "initial untransformed box" mean?

File: components/layout_2020/fragment_tree/box_fragment.rs [hunk 6]
```diff
@@ -238,30 +241,78 @@ impl BoxFragment {
...
+        overflow
+    }
+
+    /// <https://drafts.csswg.org/css-overflow/#unreachable-scrollable-overflow-region>
+    /// > area beyond the scroll origin in either axis is considered the unreachable scrollable overflow region
+    ///
+    /// Return the clipped the scrollable overflow based on its scroll origin, determined by overflow direction.
+    /// By default the clipping rect is the padding box of the fragment, this will coincides with the scrollport
+    /// if the fragment is a scroll container.
+    pub fn reachable_scrolling_overflow_region(
```
Comment (xiaochengh) [inline components/layout_2020/fragment_tree/box_fragment.rs:293]:
```suggestion
    pub fn reachable_scrollable_overflow_region(
```

File: components/layout_2020/fragment_tree/box_fragment.rs [hunk 7]
```diff
@@ -238,30 +241,78 @@ impl BoxFragment {
...
+        &self,
+        clipping_rect: Option<PhysicalRect<Au>>,
+    ) -> PhysicalRect<Au> {
+        let scrolling_direction = self.style.overflow_direction();
         let scrollable_overflow = self.scrollable_overflow();
-        let bottom_right = PhysicalPoint::new(
-            overflow.max_x().max(scrollable_overflow.max_x()),
-            overflow.max_y().max(scrollable_overflow.max_y()),
-        );
+        let clipping_rect = clipping_rect.unwrap_or(self.padding_rect());
```
Comment (xiaochengh) [inline components/layout_2020/fragment_tree/box_fragment.rs:299]: Let's extend `clipping_rect` as:
- if scrolling direction is rightward, extend right edge of `clipping_rect` to infinity; otherwise extend the left edge to -infinity
- extend the top/bottom edge of `clipping_rect` similarly

Then the result is simply the intersection of `scrollable_overflow` and `clipping_rect`, and we can get rid of the `(_, _)` branch

And then we can extend the tests to all combinations of directions and writing modes :)

(You may consider parameterizing the tests: https://web-platform-tests.org/writing-tests/testharness.html#variants)

**Review by xiaochengh — COMMENTED** [review]
Looking good!

File: components/layout_2020/fragment_tree/box_fragment.rs (hunk 7 above)
Comment (stevennovaryo) [reply components/layout_2020/fragment_tree/box_fragment.rs:299]: We have a test for this [scrollWidthHeight-negative-margin-002.html](https://wpt.fyi/results/css/cssom-view/scrollWidthHeight-negative-margin-002.html?label=master&product=chrome%5Bexperimental%5D&product=edge%5Bexperimental%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&product=servo&aligned&q=scrollwidthheight-negative-margin-002.html%20) but the test name contains many duplicates because `writing-mode` is not included. I guess we could make a new test that specifically test transform and does not depend on these `cssText`.

Comment (xiaochengh) [reply components/layout_2020/fragment_tree/box_fragment.rs:299]: Well, there are many ways to parameterize tests, and any of them is fine.

Comment (stevennovaryo) [reply components/layout_2020/fragment_tree/box_fragment.rs]: Added a test using the same logic as [scrollWidthHeight-negative-margin-002.html](https://wpt.fyi/results/css/cssom-view/scrollWidthHeight-negative-margin-002.html?label=master&product=chrome%5Bexperimental%5D&product=edge%5Bexperimental%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&product=servo&aligned&q=scrollwidthheight-negative-margin-002.html%20), but using transform (instead of margin), and limit the scope of the test.

**Review by xiaochengh — COMMENTED** [review]
LGTM, thanks!

Could a maintainer review this now?

File: tests/wpt/tests/css/css-overflow/scrollable-overflow-transform-005.html [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    width: 100px;
+    height: 100px;
+  }
+</style>
+<body onload="checkLayout('.container')">
+
+  <div class="container" data-expected-scroll-width="100" data-expected-scroll-height="100">
+    <div style="transform: translate(-200px, -200px);" class="element">
+    </div>
+  </div>
```
Comment (Loirooriol) [inline tests/wpt/tests/css/css-overflow/scrollable-overflow-transform-005.html:35]: https://drafts.csswg.org/cssom-view/#scrolling-area isn't clear about taking into account the transforms of the contents. Probably it should refer to https://drafts.csswg.org/css-overflow-3/#scrollable?

I think you should file an issue to the CSSWG, an link to it in the test with `<link rel="help"`

WebKit has the desired behavior, but is still failing the test when using classical scrollbars. If the scrollbar is 15px wide, then `scrollWidth` will be `85px`. I guess you can address this with `scrollbar-width: none`.

File: components/layout/style_ext.rs [hunk 9]
```diff
@@ -961,6 +972,24 @@ impl ComputedValuesExt for ComputedValues {
         };
         has_percentage(box_offsets.block_start) || has_percentage(box_offsets.block_end)
     }
+
+    // <https://drafts.csswg.org/cssom-view/#overflow-directions>
+    fn overflow_direction(&self) -> OverflowDirection {
+        let writing_mode: WritingMode = self.writing_mode;
+        let inline_end_direction = writing_mode.inline_end_physical_side();
+        let block_end_direction = writing_mode.block_end_physical_side();
```
Comment (Loirooriol) [inline components/layout/style_ext.rs:980]: `: WritingMode` seems unnecessary. And the variable too.

```suggestion
        let inline_end_direction = self.writing_mode.inline_end_physical_side();
        let block_end_direction = self.writing_mode.block_end_physical_side();
```

File: components/layout/display_list/mod.rs [hunk 10]
```diff
@@ -39,7 +39,6 @@ use webrender_api::{
 use wr::units::LayoutVector2D;
 
 use crate::context::{LayoutContext, ResolvedImage};
-use crate::display_list::conversions::ToWebRender;
```
Comment (Loirooriol) [inline components/layout/display_list/mod.rs:42]: Maybe just make this one public instead of switching to a wildcard?

**Review by Loirooriol — CHANGES_REQUESTED** [review]

File: components/layout/display_list/mod.rs (hunk 10 above)
Comment (stevennovaryo) [reply components/layout/display_list/mod.rs:42]: Done, thanks!

File: components/layout/style_ext.rs (hunk 9 above)
Comment (stevennovaryo) [reply components/layout/style_ext.rs]: Fixed, thanks!

File: tests/wpt/tests/css/css-overflow/scrollable-overflow-transform-005.html (hunk 8 above)
Comment (stevennovaryo) [reply tests/wpt/tests/css/css-overflow/scrollable-overflow-transform-005.html]: I am not sure about what is [scrolling-area](https://drafts.csswg.org/cssom-view/#scrolling-area) in this context, I guess the `<link rel="help"` already referring to https://drafts.csswg.org/css-overflow-3/#scrollable?

Noted, I will file an issue soon. Marked the ones with unusual behavior as tentative.

Fixed scrollbar width, thank you for the review!

**Comment by stevennovaryo** [comment]
Rebased, with the recent changes modify a bit of scrolling area logic there are small changes there too.

**Review by Loirooriol — APPROVED** [review]
Sorry for the slow review!

**Review by xiaochengh — APPROVED** [review]

**Comment by stevennovaryo** [comment]
Seems like DCO check is stuck, could we rerun the check?

**Comment by jdm** [comment]
I have not found a way to rerun a missing DCO check without force pushing a new revision or pressing the "Update branch" button to merge the latest changes from main.

**Comment by stevennovaryo** [comment]
Rebased, tecent refactoring modified a function utilized in this changes. Kindly check the PR.

**Comment by stevennovaryo** [comment]
Sorry, I missed several additional WPT passes.

File: tests/wpt/meta/css/cssom-view/scrollWidthHeight-negative-margin-002.html.ini [hunk 11]
```diff
@@ -121,97 +121,97 @@
     expected: FAIL
 
   [scrollWidth with negative margins: display: flex; overflow: visible; direction: ltr; flex-direction: row;]
-    expected: FAIL
+    expected: [FAIL, PASS]
```
Comment (Loirooriol) [inline tests/wpt/meta/css/cssom-view/scrollWidthHeight-negative-margin-002.html.ini:124]: Wait, why are the results are unreliable?

Comment (stevennovaryo) [reply tests/wpt/meta/css/cssom-view/scrollWidthHeight-negative-margin-002.html.ini:124]: The test uses `cssText` API, and because some CSS property is not enabled, the name contains some duplicates. I have added `tests/wpt/tests/css/css-overflow/scrollable-overflow-transform-unreachable-region.html` test which test quite similar case without  this dependency.
